library-best-practices/JLBP-6.md

```diff
@@ -0,0 +1,141 @@
+package) is the least disruptive and the only option that won't lead to diamond
+dependency conflicts, so it should be the generally preferred approach. However,
+given that a full transition from one major version to the next requires
+updating all import statements regardless of whether the corresponding classes
+have changed, there is potentially a very high cost to this option, which should
+be weighed carefully. Case 1 (neither renaming the maven coordinates nor the
+Java package) doesn't have this code updating cost, and the fewer the breakages,
+the easier the transition is between major versions. Basically, the cost of
+diamond dependency conflicts has to be weighed against the cost of updating
+import statements everywhere the library is used.
+
+Let's take examples from two extremes.
+
+1. A library with 10,000 classes, and 1 function is deprecated between major
+   version 1 to major version 2, and it is used in one place in a large
+   dependency tree.
+
+In this case, moving 10,000 classes to a new package in a large dependency tree
+would be a very expensive endeavor. In contrast, updating the one place where
+the old function is used to use the new function instead would be considerably
+less work and could be rolled out much more quickly. In this scenario, it is
+clearly superior to keep the same maven coordinates and Java package.
+
+3. A library with 1,000 classes, and a large refactoring touching 750 of the
+   classes is done between major version 1 and major version 2.
+
+In this case, changing consuming code would be a large undertaking, and it's not
+certain that all consuming code would feel it's worth it to migrate to the new
+major version. If the library author decided to opt for keeping the same maven
+coordinates and same Java package, the ecosystem would need to bifurcate in
+order to handle code paths requiring one versus the other major version. Either
+the ecosystem would generally opt to retain the old major version, or there
+would be an extended period of difficult diamond dependency conflicts before
+everyone had transitioned. In this scenario, it is clearly superior to rename
+both the maven coordinates and Java package.
+
+Given these considerations, the rule to follow can be summed up as:
+
+*When making breaking changes between major versions, prefer to rename both the
+maven coordinates and the Java package. However, if the ratio of surface
+breakage to library size is sufficiently low and the number of usage points of
+breaking code is sufficiently low, keeping the same maven coordinates and Java
+package can be preferred instead. Under no circumstance should only one of the
+two be renamed without the other being renamed also.*
+
+Example renames in open source 
+[Square has established it as a policy](http://jakewharton.com/java-interoperability-policy-for-major-version-updates/)
+for its Java libraries (examples include OkHttp and Retrofit). Numerous open
```

Comment:
I feel like far more have not renamed the packages.

Reply:
This is probably true, but I didn't collect the evidence. Do you have some examples that I can put in? (Guava is obviously one)